### ISSUE TITLE
Webui update fails but cli upgrade works : issue with check.sh

### ISSUE THREAD

**Issue by bluesangel** [opnsense/core#4399]
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

[X] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md

[X] I have searched the existing issues and I'm convinced that mine is new.

**Describe the bug**
I installed 20.7 performed upgrade to 20.7.1 20.7.2 20.7.3 
At some point (20.7.1 I believe) Webui update does no longer work - but update can be done through CLI command opnsense-update

**To Reproduce**
1. Go to https://xxxxxx/ui/core/firmware#updates
2. Click on 'Check for updates'
3. Scroll down to '....' / NA
4. See error : 
Timeout while connecting to the selected mirror.

**Expected behavior**
Successful check & download of updated packages 

**Screenshots**

**Relevant log files**

2020-10-06T11:10:19 | configd.py[22885] | [5c851f81-fc24-49ef-9f98-79bd0c101366] retrieve upgrade progress status
-- | -- | --
2020-10-06T11:10:19 | configd.py[22885] | [e012707e-f2b6-4d1c-83d3-21d4554be0b7] Retrieving changelog index
2020-10-06T11:10:19 | configd.py[22885] | [ea579575-791b-4cb2-904e-46c38647830b] view local packages
2020-10-06T11:10:19 | configd.py[22885] | [70a8f05f-0c80-4a5e-801e-54107bf2973f]  Script action failed with Command 'pkg rquery  "%n\|\|\|%v\|\|\|%c\|\|\|%sh\|\|\|0\|\|\|%L\|\|\|%R\|\|\|%o"' returned non-zero exit status  3. at Traceback (most recent call last):   File  "/usr/local/opnsense/service/modules/processhandler.py", line 479, in  execute     stdout=output_stream, stderr=error_stream)   File  "/usr/local/lib/python3.7/subprocess.py", line 363, in check_call      raise CalledProcessError(retcode, cmd) subprocess.CalledProcessError:  Command 'pkg rquery "%n\|\|\|%v\|\|\|%c\|\|\|%sh\|\|\|0\|\|\|%L\|\|\|%R\|\|\|%o"' returned  non-zero exit status 3.
2020-10-06T11:10:14 | configd.py[22885] | [70a8f05f-0c80-4a5e-801e-54107bf2973f] view remote packages
2020-10-06T11:09:38 | configd.py[22885] | [e4b63cbe-dbe8-4f3a-995b-a54449fd3027] retrieve firmware update status
2020-10-06T11:09:37 | configd.py[22885] | [cc2c9c12-6430-4d47-814a-60f14eb8acb3] Fetching changelog from remote
2020-10-06T11:09:37 | configd.py[22885] | [b2f80b23-0b40-4125-b137-8ea25fa311dd] Probing release type name

**Additional context**
CLI command ouput 
#  opnsense-update
Updating OPNsense repository catalogue...
Fetching meta.conf: 100%    163 B   0.2kB/s    00:01    
Fetching packagesite.txz: 100%  193 KiB 197.2kB/s    00:01    
Processing entries: 100%
OPNsense repository update completed. 721 packages processed.
All repositories are up to date.
Updating OPNsense repository catalogue...
OPNsense repository is up to date.
All repositories are up to date.
Checking for upgrades (60 candidates): 100%
Processing candidates (60 candidates): 100%
Checking integrity... done (0 conflicting)
Your packages are up to date.
Checking integrity... done (0 conflicting)
Nothing to do.
Nothing to do.
Your system is up to date.

pkg commands are ok / pkg rquery  work fine through CLI : 

Webui AUDIT reports no issue : 
***GOT REQUEST TO AUDIT HEALTH***
>>> Check installed kernel version
Version 20.7.3 is correct.
>>> Check for missing or altered kernel files
No problems detected.
>>> Check installed base version
Version 20.7.3 is correct.
>>> Check for missing or altered base files
No problems detected.
>>> Check for and install missing package dependencies
Checking all packages: .......... done
>>> Check for missing or altered package files
Checking all packages: .......... done
>>> Check for core packages consistency
Checking core packages: .
beep-1.0_1 has no upstream equivalent
Checking core packages: .
bsdinstaller-20.7 has no upstream equivalent
Checking core packages: .
ca_root_nss-3.57 has no upstream equivalent
Checking core packages: .
choparp-20150613 has no upstream equivalent
Checking core packages: .
cpustats-0.1 has no upstream equivalent
Checking core packages: .
dhcp6c-20200512 has no upstream equivalent
Checking core packages: .
dhcpleases-0.2 has no upstream equivalent
Checking core packages: .
dnsmasq-2.82,1 has no upstream equivalent
Checking core packages: .
dpinger-3.0 has no upstream equivalent
Checking core packages: .
expiretable-0.6_1 has no upstream equivalent
Checking core packages: .
filterlog-0.3 has no upstream equivalent
Checking core packages: .
flock-2.30.1 has no upstream equivalent
Checking core packages: .
flowd-0.9.1_3 has no upstream equivalent
Checking core packages: .
hostapd-2.9_2 has no upstream equivalent
Checking core packages: .
ifinfo-10.1 has no upstream equivalent
Checking core packages: .
isc-dhcp44-relay-4.4.2_1 has no upstream equivalent
Checking core packages: .
isc-dhcp44-server-4.4.2_1 has no upstream equivalent
Checking core packages: .
lighttpd-1.4.55_1 has no upstream equivalent
Checking core packages: .
monit-5.27.0 has no upstream equivalent
Checking core packages: .
mpd5-5.9 has no upstream equivalent
Checking core packages: .
ntp-4.2.8p15 has no upstream equivalent
Checking core packages: .
openssh-portable-8.2.p1_1,1 has no upstream equivalent
Checking core packages: .
openvpn-2.4.9_3 has no upstream equivalent
Checking core packages: .
opnsense-20.7.3 has no upstream equivalent
Checking core packages: .
opnsense-lang-20.1.4 has no upstream equivalent
Checking core packages: .
opnsense-update-20.7.3 has no upstream equivalent
Checking core packages: .
pam_opnsense-19.1.3 has no upstream equivalent
Checking core packages: .
pftop-0.7_9 has no upstream equivalent
Checking core packages: .
php73-ctype-7.3.22 has no upstream equivalent
Checking core packages: .
php73-curl-7.3.22 has no upstream equivalent
Checking core packages: .
php73-dom-7.3.22 has no upstream equivalent
Checking core packages: .
php73-filter-7.3.22 has no upstream equivalent
Checking core packages: .
php73-gettext-7.3.22 has no upstream equivalent
Checking core packages: .
php73-google-api-php-client-2.4.0 has no upstream equivalent
Checking core packages: .
php73-hash-7.3.22 has no upstream equivalent
Checking core packages: .
php73-json-7.3.22 has no upstream equivalent
Checking core packages: .
php73-ldap-7.3.22 has no upstream equivalent
Checking core packages: .
php73-openssl-7.3.22 has no upstream equivalent
Checking core packages: .
php73-pdo-7.3.22 has no upstream equivalent
Checking core packages: .
php73-pecl-radius-1.4.0.b1 has no upstream equivalent
Checking core packages: .
php73-phalcon-3.4.5 has no upstream equivalent
Checking core packages: .
php73-phpseclib-2.0.23 has no upstream equivalent
Checking core packages: .
php73-session-7.3.22 has no upstream equivalent
Checking core packages: .
php73-simplexml-7.3.22 has no upstream equivalent
Checking core packages: .
php73-sockets-7.3.22 has no upstream equivalent
Checking core packages: .
php73-sqlite3-7.3.22 has no upstream equivalent
Checking core packages: .
php73-xml-7.3.22 has no upstream equivalent
Checking core packages: .
php73-zlib-7.3.22 has no upstream equivalent
Checking core packages: .
pkg-1.15.6 has no upstream equivalent
Checking core packages: .
py37-Jinja2-2.10.1 has no upstream equivalent
Checking core packages: .
py37-dnspython-1.16.0 has no upstream equivalent
Checking core packages: .
py37-netaddr-0.7.19_1 has no upstream equivalent
Checking core packages: .
py37-requests-2.22.0 has no upstream equivalent
Checking core packages: .
py37-sqlite3-3.7.9_7 has no upstream equivalent
Checking core packages: .
py37-ujson-3.0.0 has no upstream equivalent
Checking core packages: .
radvd-2.18_1 has no upstream equivalent
Checking core packages: .
rate-0.9_2 has no upstream equivalent
Checking core packages: .
rrdtool-1.7.2_4 has no upstream equivalent
Checking core packages: .
samplicator-1.3.8.r1_1 has no upstream equivalent
Checking core packages: .
squid-4.13 has no upstream equivalent
Checking core packages: .
strongswan-5.8.4_1 has no upstream equivalent
Checking core packages: .
sudo-1.9.2 has no upstream equivalent
Checking core packages: .
suricata-5.0.3 has no upstream equivalent
Checking core packages: .
syslog-ng327-3.27.1_2 has no upstream equivalent
Checking core packages: .
syslogd-12.1 has no upstream equivalent
Checking core packages: .
unbound-1.11.0 has no upstream equivalent
Checking core packages: .
wpa_supplicant-2.9_7 has no upstream equivalent
Checking core packages: .
zip-3.0_1 has no upstream equivalent
***DONE***

**Environment**
OPNsense 20.7.3-amd64
FreeBSD 12.1-RELEASE-p10-HBSD
OpenSSL 1.1.1g  21 Apr 2020

**Comment by bluesangel** [comment]
I performed additional investigations issues are related to **pkg** calls used in `check.sh `: 

# sh -x  ./scripts/firmware/check.sh
+ base_to_reboot=''
+ connection=error
+ download_size=''
+ itemcount=0
+ kernel_to_reboot=''
+ last_check=unknown
+ linecount=0
+ packages_downgraded=''
+ packages_new=''
+ packages_upgraded=''
+ pkg_selected=''
+ pkg_upgraded=''
+ repository=error
+ timeout_update=30
+ timeout_upgrade=60
+ updates=''
+ upgrade_needs_reboot=0
+ pidfile=/tmp/pkg_update.pid
+ outfile=/tmp/pkg_update.out
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=''
+ [ -z '' ]
+ timer=30
+ pkg_running=started
+ :
+ daemon -p /tmp/pkg_update.pid -o /tmp/pkg_update.out pkg update -f
+ [ -n started -a 30 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 30 - 1
+ timer=29
+ [ -n 30563 -a 29 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 29 - 1
+ timer=28
+ [ -n 30563 -a 28 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 28 - 1
+ timer=27
+ [ -n 30563 -a 27 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 27 - 1
+ timer=26
+ [ -n 30563 -a 26 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 26 - 1
+ timer=25
+ [ -n 30563 -a 25 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 25 - 1
+ timer=24
+ [ -n 30563 -a 24 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 24 - 1
+ timer=23
+ [ -n 30563 -a 23 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 23 - 1
+ timer=22
+ [ -n 30563 -a 22 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 22 - 1
+ timer=21
+ [ -n 30563 -a 21 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 21 - 1
+ timer=20
+ [ -n 30563 -a 20 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 20 - 1
+ timer=19
+ [ -n 30563 -a 19 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 19 - 1
+ timer=18
+ [ -n 30563 -a 18 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 18 - 1
+ timer=17
+ [ -n 30563 -a 17 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 17 - 1
+ timer=16
+ [ -n 30563 -a 16 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 16 - 1
+ timer=15
+ [ -n 30563 -a 15 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 15 - 1
+ timer=14
+ [ -n 30563 -a 14 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 14 - 1
+ timer=13
+ [ -n 30563 -a 13 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 13 - 1
+ timer=12
+ [ -n 30563 -a 12 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 12 - 1
+ timer=11
+ [ -n 30563 -a 11 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 11 - 1
+ timer=10
+ [ -n 30563 -a 10 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 10 - 1
+ timer=9
+ [ -n 30563 -a 9 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 9 - 1
+ timer=8
+ [ -n 30563 -a 8 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 8 - 1
+ timer=7
+ [ -n 30563 -a 7 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 7 - 1
+ timer=6
+ [ -n 30563 -a 6 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 6 - 1
+ timer=5
+ [ -n 30563 -a 5 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 5 - 1
+ timer=4
+ [ -n 30563 -a 4 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 4 - 1
+ timer=3
+ [ -n 30563 -a 3 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 3 - 1
+ timer=2
+ [ -n 30563 -a 2 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 2 - 1
+ timer=1
+ [ -n 30563 -a 1 -ne 0 ]
+ sleep 1
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ expr 1 - 1
+ timer=0
+ [ -n 30563 -a 0 -ne 0 ]
+ [ 0 -eq 0 ]
+ pgrep -nF /tmp/pkg_update.pid
+ pkg_running=30563
+ [ -n 30563 ]
+ pkill -F /tmp/pkg_update.pid
+ connection=timeout
+ [ 0 -gt 0 ]
+ sed 's/"/\\&/g'
+ tr '\n' ' '
+ cat /usr/local/opnsense/firmware-message
+ upgrade_major_message=''
+ cat /usr/local/opnsense/firmware-upgrade
+ upgrade_major_version=''
+ opnsense-version -v
+ product_version=20.7.4
+ opnsense-version -n
+ product_name=opnsense
+ uname -sr
+ os_version='FreeBSD 12.1-RELEASE-p10-HBSD'
+ date
+ last_check='Mon Oct 26 19:42:02 CET 2020'
+ cat
{
        "connection":"timeout",
        "downgrade_packages":[],
        "download_size":"",
        "last_check":"Mon Oct 26 19:42:02 CET 2020",
        "new_packages":[],
        "os_version":"FreeBSD 12.1-RELEASE-p10-HBSD",
        "product_name":"opnsense",
        "product_version":"20.7.4",
        "reinstall_packages":[],
        "remove_packages":[],
        "repository":"error",
        "updates":"",
        "upgrade_major_message":"",
        "upgrade_major_version":"",
        "upgrade_needs_reboot":"0",
        "upgrade_packages":[]
}

But strangely  `pkg update -f `works fine : 
cat /tmp/pkg_update.out
Updating OPNsense repository catalogue...
Fetching meta.conf: . done
Fetching packagesite.txz: .......... done
Processing entries: .......... done
OPNsense repository update completed. 722 packages processed.
All repositories are up to date.

**Comment by fichtner** [comment]
This doesn't seem to be an issue "works" vs. "does not work": pkg update -f times out after 30 seconds which might indicate an issue. Wether the commands works or not is irrelevant and you should measure how long it takes first before starting to investigate why it takes longer than 30 seconds most probably. Sounds like a DNS timeout (IPv6 first fails).

**Comment by bluesangel** [comment]
Danke Franco. 
I don't think there is a DNS timeout issue / IPv6 first problem. 
I'm using the OPNsense Ghz small device bought in 2015. 

pkg update -f (even when using IPv4) takes more than 30 seconds on this hardware: 
`# time pkg update -f
Updating OPNsense repository catalogue...
Fetching meta.conf: 100%    163 B   0.2kB/s    00:01    
Fetching packagesite.txz: 100%  195 KiB 199.6kB/s    00:01    
Processing entries: 100%
OPNsense repository update completed. 722 packages processed.
All repositories are up to date.
       33.32 real         0.80 user         0.16 sys
# time pkg update -f
Updating OPNsense repository catalogue...
Fetching meta.conf: 100%    163 B   0.2kB/s    00:01    
Fetching packagesite.txz: 100%  195 KiB 199.6kB/s    00:01    
Processing entries: 100%
OPNsense repository update completed. 722 packages processed.
All repositories are up to date.
       48.31 real         0.79 user         0.15 sys
# time pkg -4 update -f
Updating OPNsense repository catalogue...
Fetching meta.conf: 100%    163 B   0.2kB/s    00:01    
Fetching packagesite.txz: 100%  195 KiB 199.6kB/s    00:01    
Processing entries: 100%
OPNsense repository update completed. 722 packages processed.
All repositories are up to date.
       38.37 real         0.79 user         0.16 sys
` 
I increased **timeout_update** in **check.sh** to **50** seconds and it solves the issue for the cli command check.sh : 
```
# ./check.sh
{
        "connection":"ok",
        "downgrade_packages":[],
        "download_size":"",
        "last_check":"Tue Nov  3 16:08:54 CET 2020",
        "new_packages":[],
        "os_version":"FreeBSD 12.1-RELEASE-p10-HBSD",
        "product_name":"opnsense",
        "product_version":"20.7.4",
        "reinstall_packages":[],
        "remove_packages":[],
        "repository":"ok",
        "updates":"0",
        "upgrade_major_message":"",
        "upgrade_major_version":"",
        "upgrade_needs_reboot":"0",
        "upgrade_packages":[]
}
# 
```

**Comment by fichtner** [comment]
Ok, can you check manually how long "Processing entries:" update takes? It shoudn't take more than 2-3 seconds. What hardware are you using?

**Comment by fichtner** [comment]
PS: SD card maybe?

**Comment by bluesangel** [comment]
Device is https://www.applianceshop.eu/security-appliances/security-appliances-desktop-and-wallmountable/opnsense-small-ghz.html 
Yes it uses SD card. 
CPU is  AMD T40E APU 1Ghz Dual Core
PkgRepo: extracting signature of repo in a sandbox seems to be the part taking the most time. 

```
# (while :; do echo 10; sleep 10; done) & 
# 10
# time pkg -4 -d update -f 
DBG(1)[98110]> pkg initialized
Updating OPNsense repository catalogue...
DBG(1)[98110]> PkgRepo: verifying update for OPNsense
DBG(1)[98110]> Pkgrepo, begin update of '/var/db/pkg/repo-OPNsense.sqlite'
DBG(1)[98110]> Request to fetch pkg+https://pkg.opnsense.org/FreeBSD:12:amd64/20.7/latest/meta.conf
DBG(1)[98110]> opening libfetch fetcher
DBG(1)[98110]> Fetch > libfetch: connecting
DBG(1)[98110]> Fetch: fetching from: https://pkg.opnsense.org/FreeBSD:12:amd64/20.7/latest/meta.conf with opts "i4"
DBG(1)[98110]> Fetch: fetcher chosen: https
Fetching meta.conf: 100%    163 B   0.2kB/s    00:01    
DBG(1)[98110]> Request to fetch pkg+https://pkg.opnsense.org/FreeBSD:12:amd64/20.7/latest/packagesite.txz
DBG(1)[98110]> opening libfetch fetcher
DBG(1)[98110]> Fetch > libfetch: connecting
DBG(1)[98110]> Fetch: fetching from: https://pkg.opnsense.org/FreeBSD:12:amd64/20.7/latest/packagesite.txz with opts "i4"
DBG(1)[98110]> Fetch: fetcher chosen: https
Fetching packagesite.txz: 100%  195 KiB 199.6kB/s    00:01    
DBG(1)[98110]> PkgRepo: extracting packagesite.yaml of repo OPNsense
DBG(1)[99433]> PkgRepo: extracting signature of repo in a sandbox
10
10
10
10
DBG(1)[98110]> Pkgrepo, reading new packagesite.yaml for '/var/db/pkg/repo-OPNsense.sqlite'
Processing entries:  97%10
Processing entries: 100%
OPNsense repository update completed. 722 packages processed.
All repositories are up to date.
       44.48 real         0.88 user         0.19 sys
# 10

```

**Comment by fichtner** [comment]
I'm going to fix this via #4500, but it will take time to mature as firmware updates are mission critical component. :)
